File: update_sitemap.js
#!/usr/bin/env node

/**
 * Sitemap Update Script for Milford India Spice Website
 * Node.js version - updates sitemap.xml with current pages and timestamps
 */

const fs = require('fs');
const path = require('path');

class SitemapUpdater {
    constructor(baseUrl = 'https://milfordiindiaspice.com', sitemapPath = 'sitemap.xml') {
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
        this.sitemapPath = sitemapPath;
        this.pages = [];
    }

    scanHtmlFiles() {
        console.log('🔍 Scanning for HTML files...');
        
        // Define page configuration
        const pageConfig = {
            'index.html': { priority: '1.0', changefreq: 'weekly', url: '/' },
            'menu.html': { priority: '0.9', changefreq: 'weekly' }
        };

        // Get all HTML files in current directory
        const files = fs.readdirSync('.').filter(file => file.endsWith('.html'));
        
        files.forEach(file => {
            if (pageConfig[file]) {
                const config = pageConfig[file];
                const url = config.url || `/${file}`;
                
                this.pages.push({
                    loc: `${this.baseUrl}${url}`,
                    lastmod: new Date().toISOString(),
                    changefreq: config.changefreq,
                    priority: config.priority
                });
                
                console.log(`  ✓ Added: ${file} -> ${this.baseUrl}${url}`);
            }
        });

        console.log(`📄 Found ${this.pages.length} pages to include in sitemap`);
    }

    generateSitemapXml() {
        let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
        xml += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n';

        this.pages.forEach(page => {
            xml += '  <url>\n';
            xml += `    <loc>${page.loc}</loc>\n`;
            xml += `    <lastmod>${page.lastmod}</lastmod>\n`;
            xml += `    <changefreq>${page.changefreq}</changefreq>\n`;
            xml += `    <priority>${page.priority}</priority>\n`;
            xml += '  </url>\n';
        });

        xml += '</urlset>\n';
        return xml;
    }

    backupExistingSitemap() {
        if (fs.existsSync(this.sitemapPath)) {
            const backupPath = `${this.sitemapPath}.backup`;
            fs.copyFileSync(this.sitemapPath, backupPath);
            console.log(`💾 Backed up existing sitemap to ${backupPath}`);
            return true;
        }
        return false;
    }

    updateSitemap() {
        console.log('🚀 Starting sitemap update...');
        
        // Scan for HTML files
        this.scanHtmlFiles();

        if (this.pages.length === 0) {
            console.log('❌ No HTML pages found to include in sitemap');
            return false;
        }

        // Backup existing sitemap
        this.backupExistingSitemap();

        // Generate new sitemap
        const sitemapXml = this.generateSitemapXml();

        // Write to file
        try {
            fs.writeFileSync(this.sitemapPath, sitemapXml, 'utf8');
            console.log(`✅ Sitemap updated successfully: ${this.sitemapPath}`);
            console.log(`📊 Added ${this.pages.length} pages to sitemap`);
            return true;
        } catch (error) {
            console.error('❌ Error writing sitemap:', error.message);
            return false;
        }
    }

    validateSitemap() {
        console.log('🔍 Validating sitemap...');
        
        try {
            const content = fs.readFileSync(this.sitemapPath, 'utf8');
            
            // Basic XML validation
            if (!content.includes('<?xml version="1.0"')) {
                console.log('⚠️  Warning: Missing XML declaration');
                return false;
            }
            
            if (!content.includes('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">')) {
                console.log('⚠️  Warning: Missing or incorrect urlset declaration');
                return false;
            }

            // Count URLs
            const urlMatches = content.match(/<url>/g);
            const urlCount = urlMatches ? urlMatches.length : 0;
            
            console.log(`✅ Sitemap validation passed: ${urlCount} URLs found`);
            return true;
            
        } catch (error) {
            console.error('❌ Sitemap validation failed:', error.message);
            return false;
        }
    }

    printSummary() {
        console.log('\n📋 Sitemap Update Summary');
        console.log('========================');
        console.log(`🌐 Website: ${this.baseUrl}`);
        console.log(`📄 Sitemap file: ${this.sitemapPath}`);
        console.log(`📊 Total URLs: ${this.pages.length}`);
        console.log(`🕒 Updated: ${new Date().toLocaleString()}`);
        
        console.log('\n📌 Next Steps:');
        console.log('1. Test sitemap URL: ' + `${this.baseUrl}/sitemap.xml`);
        console.log('2. Submit to Google Search Console');
        console.log('3. Verify robots.txt references the sitemap');
        console.log('4. Use online sitemap validators to double-check');
    }
}

// Main execution
function main() {
    console.log('🌐 Milford India Spice - Sitemap Updater');
    console.log('========================================\n');

    const updater = new SitemapUpdater();
    
    const success = updater.updateSitemap();
    
    if (success) {
        updater.validateSitemap();
        updater.printSummary();
    } else {
        console.log('\n❌ Failed to update sitemap');
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = SitemapUpdater;

File: update_sitemap.py
#!/usr/bin/env python3
"""
Sitemap Update Script for Milford India Spice Website
This script automatically updates the sitemap.xml file with current pages and timestamps.
"""

import os
import xml.etree.ElementTree as ET
from datetime import datetime
import glob


class SitemapUpdater:
    def __init__(self, base_url="https://milfordiindiaspice.com", sitemap_path="sitemap.xml"):
        self.base_url = base_url.rstrip('/')
        self.sitemap_path = sitemap_path
        self.pages = []
        
    def scan_html_files(self):
        """Scan the directory for HTML files to include in sitemap."""
        html_files = glob.glob("*.html")
        
        # Define page priorities and change frequencies
        page_config = {
            "index.html": {"priority": "1.0", "changefreq": "weekly", "url": "/"},
            "menu.html": {"priority": "0.9", "changefreq": "weekly"},
        }
        
        for html_file in html_files:
            if html_file in page_config:
                config = page_config[html_file]
                url = config.get("url", f"/{html_file}")
                self.pages.append({
                    "loc": f"{self.base_url}{url}",
                    "lastmod": datetime.now().strftime("%Y-%m-%dT%H:%M:%S+00:00"),
                    "changefreq": config["changefreq"],
                    "priority": config["priority"]
                })
    
    def create_sitemap_xml(self):
        """Create the sitemap XML structure."""
        # Create root element
        urlset = ET.Element("urlset")
        urlset.set("xmlns", "http://www.sitemaps.org/schemas/sitemap/0.9")
        
        # Add each page
        for page in self.pages:
            url_elem = ET.SubElement(urlset, "url")
            
            loc_elem = ET.SubElement(url_elem, "loc")
            loc_elem.text = page["loc"]
            
            lastmod_elem = ET.SubElement(url_elem, "lastmod")
            lastmod_elem.text = page["lastmod"]
            
            changefreq_elem = ET.SubElement(url_elem, "changefreq")
            changefreq_elem.text = page["changefreq"]
            
            priority_elem = ET.SubElement(url_elem, "priority")
            priority_elem.text = page["priority"]
        
        return urlset
    
    def format_xml(self, elem):
        """Format XML with proper indentation."""
        from xml.dom import minidom
        rough_string = ET.tostring(elem, 'utf-8')
        reparsed = minidom.parseString(rough_string)
        return reparsed.toprettyxml(indent="  ")[23:]  # Remove first line
    
    def update_sitemap(self):
        """Update the sitemap.xml file."""
        # Scan for HTML files
        self.scan_html_files()
        
        if not self.pages:
            print("No HTML pages found to include in sitemap.")
            return False
        
        # Create XML structure
        urlset = self.create_sitemap_xml()
        
        # Format and write to file
        xml_content = '<?xml version="1.0" encoding="UTF-8"?>\n' + self.format_xml(urlset)
        
        # Backup existing sitemap if it exists
        if os.path.exists(self.sitemap_path):
            backup_path = f"{self.sitemap_path}.backup"
            os.rename(self.sitemap_path, backup_path)
            print(f"Backed up existing sitemap to {backup_path}")
        
        # Write new sitemap
        with open(self.sitemap_path, 'w', encoding='utf-8') as f:
            f.write(xml_content)
        
        print(f"Sitemap updated successfully: {self.sitemap_path}")
        print(f"Added {len(self.pages)} pages to sitemap")
        
        return True
    
    def validate_sitemap(self):
        """Basic validation of the sitemap XML."""
        try:
            tree = ET.parse(self.sitemap_path)
            root = tree.getroot()
            
            if root.tag != "{http://www.sitemaps.org/schemas/sitemap/0.9}urlset":
                print("Warning: Invalid root element in sitemap")
                return False
            
            urls = root.findall(".//{http://www.sitemaps.org/schemas/sitemap/0.9}url")
            print(f"Sitemap validation passed: {len(urls)} URLs found")
            return True
            
        except ET.ParseError as e:
            print(f"Sitemap validation failed: {e}")
            return False


def main():
    """Main function to update the sitemap."""
    updater = SitemapUpdater()
    
    print("Updating sitemap...")
    success = updater.update_sitemap()
    
    if success:
        print("Validating sitemap...")
        updater.validate_sitemap()
        print("\nSitemap update complete!")
        print("Remember to:")
        print("1. Submit the updated sitemap to Google Search Console")
        print("2. Check robots.txt to ensure sitemap is referenced")
        print("3. Test sitemap URL accessibility")
    else:
        print("Failed to update sitemap.")


if __name__ == "__main__":
    main()